From f64d351db142f2427c483537e5f358a037544f04 Mon Sep 17 00:00:00 2001
From: Phil Elwell <pelwell@broadcom.com>
Date: Wed, 20 Mar 2013 09:58:21 +0000
Subject: [PATCH 098/106] SW-11786: Fix USE_AFTER_FREE defect (61220)

In the error paths of parse_open, only unlock the service if it is not
NULL. In the fail path, NULL the service after it is unlocked to
prevent a double unlock.
---
 drivers/misc/vc04_services/interface/vchiq_arm/vchiq_core.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/drivers/misc/vc04_services/interface/vchiq_arm/vchiq_core.c b/drivers/misc/vc04_services/interface/vchiq_arm/vchiq_core.c
index 9f66704..f35ed4f 100644
--- a/drivers/misc/vc04_services/interface/vchiq_arm/vchiq_core.c
+++ b/drivers/misc/vc04_services/interface/vchiq_arm/vchiq_core.c
@@ -1492,6 +1492,7 @@ parse_open(VCHIQ_STATE_T *state, VCHIQ_HEADER_T *header)
 					version, version_min);
 				vchiq_loud_error_footer();
 				unlock_service(service);
+				service = NULL;
 				goto fail_open;
 			}
 			service->peer_version = version;
@@ -1557,7 +1558,8 @@ fail_open:
 	return 1;
 
 bail_not_ready:
-	unlock_service(service);
+	if (service)
+		unlock_service(service);
 
 	return 0;
 }
-- 
1.8.4.2

